nixos/modules/services/networking/rosenpass.nix

```diff
@@ -0,0 +1,178 @@
+          };
+        };
+      };
+    in
+      mkOption {
+        type = listOf peer;
+        description = "List of peers to exchange keys with.";
+        default = [];
+      };
+
+    extraConfig = mkOption {
+      type = attrs;
+      description = ''
+        Extra configuration to be merged with the generated Rosenpass configuration file.
+      '';
+      default = {};
+    };
+  };
+
+  config = mkIf cfg.enable {
+    warnings = let
+      netdevsList = attrValues config.systemd.network.netdevs;
+      publicKeyInNetdevs = peer: any (netdev: any (publicKeyInWireguardPeers peer) netdev.wireguardPeers) netdevsList;
+      publicKeyInWireguardPeers = peer: x: x.wireguardPeerConfig ? PublicKey && x.wireguardPeerConfig.PublicKey == peer.wireguard.publicKey;
+
+      # NOTE: In the message below, we tried to refer to
+      #   options.systemd.network.netdevs."<name>".wireguardPeers.*.PublicKey
+      # directly, but don't know how to traverse "<name>" and * in this path.
+      warningMsg = peer: "It appears that you have configured a Rosenpass peer with the Wireguard public key '${peer.wireguard.publicKey}' but there is no corresponding Wireguard peer configuration in any of `${options.systemd.network.netdevs}.\"<name>\".wireguardPeers.*.PublicKey`. While this may work as expected, such a scenario is unusual. Please double-check your configuration.";
+    in
+      concatMap (peer: optional (!publicKeyInNetdevs peer) (warningMsg peer)) cfg.peers;
+
+    environment.systemPackages = [cfg.package pkgs.wireguard-tools];
+
+    users.users."${cfg.user}" = {
+      isSystemUser = true;
+      createHome = false;
+      group = cfg.group;
+    };
+
+    users.groups."${cfg.group}" = {};
+
+    # NOTE: It would be possible to use systemd credentials for pqsk.
+    # <https://systemd.io/CREDENTIALS/>
+    systemd.services.rosenpass = let
+      generatePeerConfig = {
+        publicKeyFile,
+        endpoint,
+        device,
+        wireguard,
+      }:
+        {
+          inherit device;
+          public_key = publicKeyFile;
+          peer = wireguard.publicKey;
+          extra_params = [];
+        }
+        // (optionalAttrs (endpoint != null) {inherit endpoint;});
+
+      generateConfig = {
+        publicKeyFile,
+        secretKeyFile,
+        listen,
+        verbosity,
+        peers,
+        ...
+      }: {
+        inherit listen verbosity;
+        public_key = publicKeyFile;
+        secret_key = secretKeyFile;
+        peers = map generatePeerConfig peers;
+      };
+      toml = pkgs.formats.toml {};
+      configFile = toml.generate "config.toml" (recursiveUpdate (generateConfig cfg) cfg.extraConfig);
+    in {
+      wantedBy = ["multi-user.target"];
+      after = ["network-online.target"];
+      path = [pkgs.wireguard-tools];
+
+      script = "${cfg.package}/bin/rosenpass exchange-config ${configFile}";
+
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
```

Comment:
Can `rosenpass` use `DynamicUser` instead of a fixed user/group?

Reply:
Probably yes.